### PR TITLE
[#9079] Set up basic REST API endpoint

### LINKED ISSUE #9079: Set up foundation for front-end migration project
To kick off [this project](https://github.com/TEAMMATES/teammates/projects/4), some foundations need to be laid out first:
- [x] Step 1: Set up Angular 6 front-end project
  - Included: lint/build/test routine (including reliable code coverage), separate dev server from the back end, hot reload in development mode, source maps, minification when building for production usage
- [x] Step 2: Set up REST-ful back end servlet, ~possibly with the help of [Jersey](https://medium.com/@tejohnso/a-minimal-java-api-with-jetty-jersey-jackson-and-gradle-9842a145815e),~ and user authentication workflow
- [x] Step 3: Migrate few pages; concurrently, set up example of how migration for each page should be done systematically.

- [ ] Step 4 (can be deferred): Set up example tests, ranging from front-end component tests (with Karma/Jasmine), API endpoint tests (with TestNG), and E2E tests (with Selenium).

### PR DESCRIPTION
**Pull request by wkurniawan07**
Part of #9079. 

This is the last setup necessary before the first few pages can be migrated.

- [x] Set up basic web API servlet with XSRF protection
  - [x] Merged backdoor servlet here as well because backdoor servlet is essentially just another gateway to back-end logic but reserved for E2E tests usage
- [x] Implemented 3-layer access control check: (1) Check request origin, (2) Check authentication, (3) Check authorization
  - [x] Backdoor key now becomes a key to bypass authentication and authorization for E2E tests. A similar key, XSRF key, is also provided to bypass request origin check. These keys must be secure and only known to app admin (just like how backdoor key has always been).
- [x] Made a very clear separation of concern between different HTTP methods `GET`, `POST`, `PUT`, `DELETE` instead of lumping everything under the same method

### CODE REVIEW

**Comment by wkurniawan07** [comment]
@xpdavid you can review the code first; I'm currently updating `design.md` and `development.md`.

File: src/main/java/teammates/ui/newcontroller/WebApiServlet.java [hunk 1]
```diff
@@ -44,7 +48,51 @@ protected void doDelete(HttpServletRequest req, HttpServletResponse resp) throws
 
     @SuppressWarnings("PMD.AvoidCatchingThrowable") // used as fallback
     private void invokeServlet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
-        // TODO
+        resp.setHeader("Strict-Transport-Security", "max-age=31536000");
```
Comment (xpdavid) [inline src/main/java/teammates/ui/newcontroller/WebApiServlet.java]: `<transport-guarantee>CONFIDENTIAL</transport-guarantee>` in `web.xml` won't do this job?

File: src/web/app/pages-student/student-page.component.ts [hunk 2]
```diff
@@ -29,7 +29,11 @@ export class StudentPageComponent implements OnInit {
       if (res.logoutUrl) {
         this.logoutUrl = `${this.backendUrl}${res.logoutUrl}`;
       }
-      this.isValidUser = res.user && res.user.isStudent;
+      if (res.user) {
+        this.isValidUser = res.user.isStudent;
+      } else {
+        window.location.href = res.studentLoginUrl;
```
Comment (xpdavid) [inline src/web/app/pages-student/student-page.component.ts]: Should use Angular service `$location`?

Comment (wkurniawan07) [reply src/web/app/pages-student/student-page.component.ts]: `$location` in an AngularJS thing, not Angular. I however will check if the [Location API](https://angular.io/api/common/Location) can be used.

File: src/web/services/http-request.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  it('should execute POST with params', () => {
+    service.post('/url', null, { key: 'value' });
+    expect(spyHttpClient.post).toHaveBeenCalledWith(`${backendUrl}/webapi/url`, null, {
+      withCredentials,
+      params: jasmine.any(Object),
+    });
+  });
+
+  it('should execute PATCH with null body by default', () => {
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/services/http-request.service.spec.ts]: Oops why did I put PATCH there 😳

File: src/main/java/teammates/ui/newcontroller/OriginCheckFilter.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+import teammates.common.util.Config;
+import teammates.common.util.CryptoHelper;
+import teammates.common.util.Url;
+
+/**
+ * Checks and validates origin of HTTP requests.
+ */
+public class OriginCheckFilter implements Filter {
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/newcontroller/OriginCheckFilter.java]: Origin check and XSRF protection are conceptually not that far apart.

File: src/main/java/teammates/ui/newcontroller/OriginCheckFilter.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        } else if (!isHttpReferrerValid(referrer, request.getRequestURL().toString())) {
+            response.setStatus(403);
+            denyAccess("Invalid HTTP referrer.", response);
+            return;
+        }
+
+        switch (request.getMethod()) {
+        case "POST":
+        case "PUT":
+        case "DELETE":
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/newcontroller/OriginCheckFilter.java]: Take a look: the `POST` and `PUT` cases fall through.

File: src/main/java/teammates/ui/newcontroller/ActionFactory.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Returns the matching {@link Action} object for the URI and method in {@code req}.
+     */
+    public Action getAction(HttpServletRequest req) throws ActionMappingException {
+        String url = req.getRequestURL().toString();
+        String method = req.getMethod();
+        log.info("URL received: [" + method + "] " + url);
+
+        String uri = req.getRequestURI();
+        if (uri.contains(";")) {
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/newcontroller/ActionFactory.java]: Not entirely sure myself, actually. This is copied from the old `ActionFactory`.

File: src/web/app/pages-student/student-page.component.ts (hunk 2 above)
Comment (wkurniawan07) [reply src/web/app/pages-student/student-page.component.ts]: Tried; Location API is not effective for navigating to absolute URLs.

File: src/main/java/teammates/ui/newcontroller/WebApiServlet.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/newcontroller/WebApiServlet.java]: I'm actually not sure what the best practice will be. `Strict-Transport-Security` header is a purist HTTP header approach (if anything I prefer using this header) while the `transport-guarantee` is more black-boxed. In fact, I'm not even sure they're doing the same thing.

**Comment by wkurniawan07** [comment]
> one action only corresponds to one HTTP method.

You realized that `teammates.ui.controller.*Action` work that way right? It was in fact very problematic; the old controller maps `GET` and `POST` to the very same action.

File: src/web/app/pages-student/student-page.component.ts (hunk 2 above)
Comment (xpdavid) [reply src/web/app/pages-student/student-page.component.ts]: Not sure if this will become a troublesome dependency when writing unit test.

**Comment by xpdavid** [comment]
:cry: Sorry, my punctuation messes up :sweat_smile:. I mean since `ActionFactory` map each action to each HTTP method, it would be more clear to put each action under a subpackage named with its mapped HTTP method (e.g. `put`, `get`, `post`).

**Comment by wkurniawan07** [comment]
> it would be more clear to put each action under a subpackage named with its mapped HTTP method

I don't really see the benefit; HTTP methods are very implementation-specific and should not be the factor for determining a class' package, and if anything it caused unnecessary dependency to/from the sub-package (e.g. from `teammates.ui.newcontroller.get` to `teammates.ui.newcontroller` if I get what you meant correctly).

**Comment by wkurniawan07** [comment]
@xpdavid Updated the design diagrams. This can be reviewed in full now.

I haven't updated the diagrams for test driver component, and by extension I also haven't updated the overall package diagram, as I'm still unsure of how the structure will be at this point of time. I'll need to wait until I actually migrate some pages to get a clear idea on those.

**Comment by wkurniawan07** [comment]
> I think I can take part in the V7 migration for logging platform

You don't have to actually. The logging project is independent of this. You should simply patch `master` branch and I'll be responsible to transfer whatever change you created there to here.

File: src/main/java/teammates/ui/newcontroller/Action.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        String regkey = req.getParameter("regkey");
+        authType = regkey == null ? AuthType.UNAUTHENTICATED : AuthType.UNREGISTERED;
+    }
+
+    /**
+     * Returns the log message in the special format used for generating the 'activity log' for the Admin.
+     */
+    public String getLogMessage() {
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/newcontroller/Action.java]: Whatever is discussed in Slack should not have any impact here. This migration has no aim for optimizing the logs.

**Pull request merged**